src/json.ts

```diff
@@ -4,9 +4,110 @@ export type JSONValue =
   | boolean
   | number
   | Array<JSONValue>
-  // undefined are filtered out in JSON.stringify so OK to allow.
-  | {[key: string]: JSONValue | undefined};
+  | JSONObject;
+
+type JSONObject = {[key: string]: JSONValue};
 
 export type ToJSON = {
-  toJSON: () => JSONValue;
+  toJSON: (key?: string) => JSONValue;
 };
+
+/**
+ * Like [[JSONValue]] but allows [[ToJSON]] at every recursive value. This
+ * matches the semantics of `JSON.stringify`.
+ */
+export type ToJSONValue =
+  | ToJSON
```

Comment:
Do we need `ToJSON` in any existing sample or anyplace else? I was imagining restricting subscribe to returning `JSONValue`. Seems this would make deepEquals easier too.

Reply:
I was even considering removing ToJSON everywhere in Replicache. I had it in places where things eventually gets fed into JSON.stringify